File: ailist/src/test/scala/me/kosik/interwalled/ailist/ListSplittingTest.scala
package me.kosik.interwalled.ailist

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class ListSplittingTest extends AnyFunSuite with Matchers {
  import AIListTestHelper._

  test("Splitting list into components test: basic splitting intervals") {
    val intervals = Array(
      Interval(  1,  11, "NIL"),
      Interval(  1,  12, "NIL"),
      Interval(  2,  13, "NIL"),
      Interval(  2,  14, "NIL"),
      Interval(  3,  15, "NIL"),

      Interval(  4,  64, "NIL"),

      Interval(  4,  12, "NIL"),
      Interval(  5,  10, "NIL"),
      Interval(  6,  12, "NIL"),
      Interval(  6,  14, "NIL"),
      Interval(  7,  18, "NIL"),

      Interval(  8,  64, "NIL"),
      Interval(  8,  62, "NIL"),
      Interval(  9,  63, "NIL"),
      Interval(  9,  61, "NIL"),
      Interval( 10,  60, "NIL"),

      Interval( 10,  33, "NIL"),
      Interval( 11,  32, "NIL"),
      Interval( 12,  33, "NIL"),
      Interval( 12,  32, "NIL"),
      Interval( 15,  33, "NIL"),
    )

    val aiList = buildList(intervals, new AIListConfig(10, 5, 5, 3))

    assert(21 == aiList.size,               "All 21 Intervals should be persisted.")
    assert( 2 == aiList.getComponentsCount, "Intervals ending with > 60 should be placed in a separate component.")
  }

  test("Splitting list into components test - no splitting intervals due to minimum component length") {
    val intervals = Array(
      Interval(  1,  11, "NIL"),
      Interval(  1,  12, "NIL"),
      Interval(  2,  13, "NIL"),
      Interval(  2,  14, "NIL"),
      Interval(  3,  15, "NIL"),

      Interval(  4,  64, "NIL"),

      Interval(  4,  12, "NIL"),
      Interval(  5,  10, "NIL"),
      Interval(  6,  12, "NIL"),
      Interval(  6,  14, "NIL"),
      Interval(  7,  18, "NIL"),

      Interval(  8,  64, "NIL"),
      Interval(  8,  62, "NIL"),
      Interval(  9,  63, "NIL"),
      Interval(  9,  61, "NIL"),
      Interval( 10,  60, "NIL"),

      Interval( 10,  33, "NIL"),
      Interval( 11,  32, "NIL"),
      Interval( 12,  33, "NIL"),
      Interval( 12,  32, "NIL"),
      Interval( 15,  33, "NIL"),
    )

    val aiList = buildList(intervals, new AIListConfig(10, 5, 5, 25))

    assert(21 == aiList.size,               "All 21 Intervals should be persisted.")
    assert( 1 == aiList.getComponentsCount, "All intervals should be placed in the same component.")
  }

  test("Splitting list into components test - no splitting intervals due to maximum component count") {
    val intervals = Array(
      Interval(  1,  11, "NIL"),
      Interval(  1,  12, "NIL"),
      Interval(  2,  13, "NIL"),
      Interval(  2,  14, "NIL"),
      Interval(  3,  15, "NIL"),

      Interval(  4,  64, "NIL"),

      Interval(  4,  12, "NIL"),
      Interval(  5,  10, "NIL"),
      Interval(  6,  12, "NIL"),
      Interval(  6,  14, "NIL"),
      Interval(  7,  18, "NIL"),

      Interval(  8,  64, "NIL"),
      Interval(  8,  62, "NIL"),
      Interval(  9,  63, "NIL"),
      Interval(  9,  61, "NIL"),
      Interval( 10,  60, "NIL"),

      Interval( 10,  33, "NIL"),
      Interval( 11,  32, "NIL"),
      Interval( 12,  33, "NIL"),
      Interval( 12,  32, "NIL"),
      Interval( 15,  33, "NIL"),
    )

    val aiList = buildList(intervals, new AIListConfig(1, 5, 5, 3))

    assert(21 == aiList.size,               "All 21 Intervals should be persisted.")
    assert( 1 == aiList.getComponentsCount, "All intervals should be placed in the same component.")
  }
}

File: ailist-core/src/test/java/me/kosik/interwalled/ailist/core/ListSplittingTest.java
package me.kosik.interwalled.ailist.core;

import me.kosik.interwalled.ailist.core.model.Configuration;
import me.kosik.interwalled.ailist.core.model.Interval;
import me.kosik.interwalled.ailist.core.benchmark.ListBuilder;
import me.kosik.interwalled.ailist.core.utils.ListUtils;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;


public class ListSplittingTest {

    final private List<Interval> basicData = Arrays.asList(
        new Interval( 1,  11),
        new Interval( 1,  12),
        new Interval( 2,  13),
        new Interval( 2,  14),
        new Interval( 3,  15),

        new Interval( 4,  64),

        new Interval( 4,  12),
        new Interval( 5,  10),
        new Interval( 6,  12),
        new Interval( 6,  14),
        new Interval( 7,  18),

        new Interval( 8,  64),
        new Interval( 8,  62),
        new Interval( 9,  63),
        new Interval( 9,  61),
        new Interval(10,  60),

        new Interval(10,  33),
        new Interval(11,  32),
        new Interval(12,  33),
        new Interval(12,  32),
        new Interval(15,  33)
    );

    @Test
    void basicSplitting() {
        final Configuration aiListConfig =
            new Configuration(6, 5, 64);

        final List<AIList> aiLists =
            ListBuilder.buildAILists(aiListConfig, basicData);

        System.out.println(aiLists);

        assertEquals(21, ListUtils.sum(aiLists));
        assertEquals(2, aiLists.size());
    }

    @Test
    void forceMaximalComponentsCount() {
        final Configuration aiListConfig =
                new Configuration(5, 5, 64);

        final List<AIList> aiLists =
            ListBuilder.buildAILists(aiListConfig, basicData);

        assertEquals(21, ListUtils.sum(aiLists));
        assertEquals(1, aiLists.size(), "All intervals should be placed in the same component.");
    }

    @Test
    void splitInitialOutlier() {
        final List<AIList> aiLists = ListBuilder.buildAILists(
            new Configuration(6, 3, 64),
            Arrays.asList(
                new Interval(0,  64),
                new Interval(0,  96),
                new Interval(0, 128),
                new Interval(1,   2),
                new Interval(2,   3),
                new Interval(3,   4),
                new Interval(4,   5),
                new Interval(5,   6),
                new Interval(6,   7),
                new Interval(7,   8),
                new Interval(8,   9),
                new Interval(9,  10)
            )
        );

        assertEquals(12, ListUtils.sum(aiLists));
        assertEquals( 2, aiLists.size(), "Outliers should be in a separate group.");
        assertEquals( 9, aiLists.get(0).intervals().length, "First component should contain most elements.");
        assertEquals( 3, aiLists.get(1).intervals().length, "Last component should store the outliers.");
    }
}
